cylc/flow/config.py

```diff
@@ -2278,6 +2283,29 @@ def load_graph(self):
         for tdef in self.taskdefs.values():
             tdef.tweak_outputs()
 
+    def check_outputs(self, terminals: Iterable[str]) -> None:
+        """Check that task outputs have been registered with tasks.
+
+        Raises: WorkflowConfigError if a custom output is not defined.
+        """
+        from cylc.flow.task_qualifiers import TASK_QUALIFIERS
+        terminal_outputs = []
+        for terminal in terminals:
+            if ':' in terminal:
+                task = re.findall(TaskID.NAME_RE, terminal)[0]
+                qualifier = re.findall(GraphParser._RE_QUAL, terminal)
+                if qualifier:
+                    qualifier_str = qualifier[0].strip(':')
+                if qualifier_str and qualifier_str not in TASK_QUALIFIERS:
+                    terminal_outputs.append((task, qualifier_str))
+
+        for task, output in terminal_outputs:
```

Comment:
Can't we just split on the `:` here?
```suggestion
        for task, output in [t.split(":") for t in terminals if ":" in t]:
```

Reply:
Not if there is a suicide trigger, or an optional output or the user has used a present tense qualifier.

Reply:
Hmm, OK, I'm not sure we should allow outputs to be declared by this sort of construct: `a => ! b:x`, but it seems we do!

Otherwise, that code block strikes me as kinda heavy-weight for what it does, but maybe it doesn't matter.

Still,
- why did you use `re.findall(...)[0]` instead of `re.match()` ?
- `strip(':')` is unnecessary (the `:` is guaranteed so just slice off the final character)?

Reply:
How about like this?
```python
2303         for task, qual in [t.split(":") for t in terminals if ":" in t]:
2304             if qual in TASK_QUALIFIERS:
2305                 continue
2306             terminal_outputs.append([task.strip("!"), qual.strip("?")])
```

Reply:
> Why use re.findall(...)[0] instead of just re.match() ?

Familiarity. Match works. :)

> I think strip(':') is unnecessary there - just slice off the final character, which is guaranteed to be a :?

First char, surely?

> How about this?

If fails a pile of tests, mostly because terminals have not in fact been well parsed (my bad) - main problem is that fail/failed are not both recognised as standard outputs.

```
tests/integration/validate/test_outputs.py::test_completion_expression_invalid[expire-optional-in-graph-but-not-used-in-completion]
tests/integration/validate/test_outputs.py::test_completion_expression_invalid[alt-compvar2]
tests/integration/validate/test_outputs.py::test_completion_expression_invalid[submit-failed used in completion expression]
tests/integration/test_optional_outputs.py::test_task_completion[2]
tests/integration/reftests/test_pre_initial.py::test_over_bracketed
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[a]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[b]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[c]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[d]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[e]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[f]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[h0]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[h1]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[i]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[j]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[k]
tests/integration/test_optional_outputs.py::test_implicit_completion_expression[l]
tests/integration/test_dbstatecheck.py::test_basic
tests/integration/test_dbstatecheck.py::test_task
tests/integration/test_dbstatecheck.py::test_point
tests/integration/test_dbstatecheck.py::test_status
tests/integration/test_dbstatecheck.py::test_output
tests/integration/test_dbstatecheck.py::test_flownum
```

Reply:
>>   I think strip(':') is unnecessary there - just slice off the final character, which is guaranteed to be a :?
>  First char, surely? And isn't strip(':') more meaningful to a future reader than some mysterious slice?

Sorry, yes, first char!

`strip(':')` presumably does some unnecessary computation if the first char is *always* a `:`.  I'd just convey the meaning like this:

`foo[1:] # strip leading ":"`

Reply:
> If fails a pile of tests, ... main problem is that fail/failed are not both recognised as standard outputs.

Actually `TASK_QUALIFIERS` does include the "present tense" standard outputs.

It's just that my untested attempt should have stripped unwanted chars before comparing with standard outputs.

This passes the tests above, and IMO is computationally simpler and easier to understand:
```python
for left, right in [t.split(":") for t in terminals if ":" in t]:
    taskname = left.strip("!")  # ignore any suicide trigger
    qualifier = right.strip("?")  # ignore any optional output markers
    if qualifier in TASK_QUALIFIERS:  # ignore any standard outputs
        continue
    terminal_outputs.append([taskname, qualifier])
```
(Maybe also replace `?` with `GraphParser.OPTIONAL` etc.)

Reply:
Even better (tested)! 
```python
2292         terminal_outputs = [
2293             (a[0].strip("!"), b)
2294             for a in (t.split(':') for t in terminals if ":" in t)
2295             if (b := a[1].strip("?")) if b not in TASK_QUALIFIERS
2296         ]
```

Reply:
Walrus invalid in Python 3.7.

Reply:
Aren't we Python 3.8+ for cylc-flow now? (Oh, the UIS?)

Reply:
Yep, cylc-flow, cylc-rose & metomi-rose still 3.7 for a while. Will bump to 3.8 (or likely 3.9) when that requirement relaxes, so no walruses just yet ( :seal: * :unicorn: ).

I am loving the commit message "remove stray walrus" though!

If you're worried about efficiency, swap `[` for `(` to turn the list comp into a generator.

Reply:
> Yep, cylc-flow, cylc-rose & metomi-rose still 3.7 for a while.

cylc-uiserver, ...

But current docs say this 🫨 :

> Via Conda (recommended
> Supported Python versions
    cylc-flow supports Python 3.8+.
    cylc-uiserver currently only supports Python 3.8-3.9.

Reply:
> But current docs say this

That is correct. We can only support Python 3.8+ on conda-forge, but we still support 3.7 via pypi.

Reply:
I'm gonna resolve this convo, with or without tusked mammals!